### PR TITLE
Abstract authentication flow into new interface

### PR DESCRIPTION
**Pull request by szh**
Note: To see implementation in the CLI, see https://github.com/cyberark/cyberark-conjur-cli/pull/410

### Desired Outcome

Create a new `AuthenticationStrategy` interface that will allow us to abstract the authentication flow. The interface should
take a `CredentialProvider` and Conjur account as inputs, as well as any other arguments that are needed to authenticate,
such as `service_id` for LDAP. The interface should have a `authenticate` method which returns a Conjur auth token. All details of authentication should be handled by the `AuthenticationStrategy` interface.
Include unit tests and update existing unit and integration tests.

### Implemented Changes

- Added `AuthenticationStrategyInterface` to abstract authentication logic
- Added `AuthnAuthenticationStrategy` which implements the interface for the default authn strategy
- Updated the Api class to use the interface to implement authentication

### Connected Issue/Story

CyberArk internal issue link: [ONYX-20429](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-20429)

#### Changelog

- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [x] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [x] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [x] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,10 +6,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
-### Deprecated
-### Removed
-### Fixed
-### Security
+- Abstract authentication flow into new `AuthenticationStrategyInterface`
+  [conjur-api-python#20](https://github.com/cyberark/conjur-api-python/pull/20)
+
+## [8.0.0] - 2022-05-25
+
+[Unreleased]: https://github.com/cyberark/cyberark-conjur-cli/compare/v8.0.0...HEAD
```
[the remark this answers is not in the document]
Comment (szh) [reply CHANGELOG.md:15]: No, good catch

[the remark this answers is not in the document]
Comment (szh) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,10 +6,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
-- The `get_server_info` method is now available in SDK. It is only supported against Conjur enterprise server 
 ### Changed
-### Deprecated
-### Removed
-### Fixed
-### Security
+- Abstract authentication flow into new `AuthenticationStrategyInterface`
+  [conjur-api-python#20](https://github.com/cyberark/conjur-api-python/pull/20)
+
+## [8.0.0] - 2022-05-25
```
Comment (doodlesbykumbi) [inline CHANGELOG.md:13]: Bit confused about this. 
1. How is it that this repo has never had any releases ?
2. 8.0.0 ?

File: conjur_api/interface/authentication_strategy_interface.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+# pylint: disable=too-few-public-methods
+class AuthenticationStrategyInterface(metaclass=abc.ABCMeta):  # pragma: no cover
+    """
+    AuthenticationStrategyInterface
+    This class is an interface that outlines a shared interface for authentication strategies
+    """
+
+    @abc.abstractmethod
+    async def authenticate(self, api_key: str, ssl_verification_data: SslVerificationMetadata) -> str:
```
Comment (doodlesbykumbi) [inline conjur_api/interface/authentication_strategy_interface.py:21]: I might be missing something on the authenticate signature on this interface. Not all authenticators use the `api_key` parameter. I think they each have their own set of parameters required to carry out authentication.

File: conjur_api/providers/authentication_strategy_factory.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+This module job is to encapsulate the creation of SSLContext in the dependent of each os environment
+"""
+
+from conjur_api.errors.errors import UnknownAuthnTypeError
+from conjur_api.interface.authentication_strategy_interface import AuthenticationStrategyInterface
+from conjur_api.models.enums.authn_types import AuthnTypes
+from conjur_api.providers.authn_authentication_strategy import AuthnAuthenticationStrategy
+
+# pylint: disable=too-few-public-methods
+def create_authentication_strategy(
```
Comment (doodlesbykumbi) [inline conjur_api/providers/authentication_strategy_factory.py:14]: This comment follows from my comments around the authenticate method on the interface.

I think each authenticator has its own set of input parameters and so I think we might need to rework this to take that into account.

**Review by doodlesbykumbi — COMMENTED** [review]
Left some comments

File: CHANGELOG.md (hunk 2 above)
Comment (szh) [reply CHANGELOG.md:13]: Basically there used to be one repo which included the python api and CLI, with the latest release being 7.1.0. Then the two were split into separate repos. So the python api repo has no releases, but is sort of starting out at 7.1.0. This PR introduces breaking changes to the API, so we're bumping it to 8.0. Meanwhile the CLI implementation of this won't have breaking changes so it'll be 7.2.

File: conjur_api/interface/authentication_strategy_interface.py (hunk 3 above)
Comment (szh) [reply conjur_api/interface/authentication_strategy_interface.py:21]: Can you give me an example of an authenticator that doesn't use an api key?

File: conjur_api/providers/authentication_strategy_factory.py (hunk 4 above)
Comment (szh) [reply conjur_api/providers/authentication_strategy_factory.py:14]: This is true. The idea would be to add in more named optional parameters to the `create_authentication_strategy` method as we add new authentication strategies.

File: CHANGELOG.md (hunk 2 above)
Comment (doodlesbykumbi) [reply CHANGELOG.md:13]: Thanks!

File: conjur_api/interface/authentication_strategy_interface.py (hunk 3 above)
Comment (jvanderhoof) [reply conjur_api/interface/authentication_strategy_interface.py:21]: I agree w/ @doodlesbykumbi. Only the default authenticator uses the API key.  All other authenticators validate the external identity and return a Conjur Auth token.

What about injecting the strategy along the lines of:
```py
credentials = CredentialsData(username=username, password=password, machine=conjur_url)
credentials_provider = SimpleCredentialsProvider()
credentials_provider.save(credentials)
del credentials

client = Client(connection_info,
                authentication_strategy=Authn(credentials_provider=credentials_provider)
                ssl_verification_mode=ssl_verification_mode)

client = Client(connection_info,
                authentication_strategy=AuthnLdap(credentials_provider=credentials_provider)
                ssl_verification_mode=ssl_verification_mode)

client = Client(connection_info,
                authentication_strategy=AuthnJwt(jwt='path-to-jwt')
                ssl_verification_mode=ssl_verification_mode)
```
If the Authentication Strategy interface includes a method (maybe `authenticate`), and returns a Conjur Auth Token, we can inject the "how" into our client.

File: conjur_api/providers/authn_authentication_strategy.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            self,
+            url: str,
+            account: str,
+            login_id: str
+    ):
+        self._url = url
+        self._account = account
+        self._login_id = login_id
+
+    async def authenticate(self, api_key: str, ssl_verification_data) -> str:
```
Comment (jvanderhoof) [inline conjur_api/providers/authn_authentication_strategy.py:30]: As mentioned above, having this method take `CredentialsData` via a `SimpleCredentialsProvider` allows the authentication to be handled here.  

As a note, we may need to add a second/alternative `CredentialsData` type to handle hostname/api key.  Users exchange a username/password for an api key, then authenticate with the username/api key combination.  Hosts authenticate with `host/<hostname>`/api key combination (note the `host` in front of the hostname).

**Review by rpothier — APPROVED** [review]
LGTM, if there are remaining items from JAson or Kumbi they can be addressed in a follow up so this can move forward.

**Pull request merged**